### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [74]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data_df = pd.read_csv(file_to_load)
purchase_data_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

## Player Count

* Display the total number of players


In [75]:
#Find the total number of players -- "SN" = screen name, use to count total #
# Use len to go into the SN's row of data and gather that total #
total_players = len(purchase_data_df["SN"].value_counts())

# Create a dataframe to display the total player count...set index to 0 
df_players = pd.DataFrame({"Total Players": total_players}, index = [0])

df_players

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [76]:
#Number of Unique Items
# Use the len().value_counts to find the total unique item numbers
unique_items = len(purchase_data_df["Item ID"].value_counts())

# Show a cleaner data display
df_unique_items= pd.DataFrame({"Number of Unique Items": unique_items}, index = [0])
# df_unique_items

In [77]:
#Average Purchase Price
average_purchase_price = purchase_data_df["Price"].mean()

# Show a cleaner data display
df_average_purchase_price = pd.DataFrame({"Average Purchase Price": average_purchase_price}, index = [0])
# df_average_purchase_price


In [78]:
# Total Number of Purchases
total_purchases = len(purchase_data_df["Price"])
total_purchases

# Show a cleaner data display
df_total_purchases = pd.DataFrame({"Total Number of Purchases": total_purchases}, index = [0])
# df_total_purchases

In [79]:
# Total Revenue
revenue = purchase_data_df["Price"].sum()
revenue

# Display with clean data
df_revenue = pd.DataFrame({"Total Revenue": revenue}, index = [0])
# df_revenue

In [80]:
#Displaying the purchasing analysis Summary Data Frame
df_purchasing_analysis = pd.DataFrame({"Number of Unique Items": unique_items,
                                      "Average Purchase Price": average_purchase_price, 
                                      "Total Number of Purchases": total_purchases,
                                      "Total Revenue": revenue}, index = [0])
#Displaying the purchasing analysis Summary Data Frame
df_purchasing_analysis = pd.DataFrame({"Number of Unique Items": unique_items,
                                      "Average Purchase Price": average_purchase_price, 
                                      "Total Number of Purchases": total_purchases,
                                      "Total Revenue": revenue}, index = [0])

df_purchasing_analysis

Number of Unique Items  Average Purchase Price  Total Number of Purchases  \
0                     179                3.050987                        780   

   Total Revenue  
0        2379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [81]:
# Create a full count to be able to divide against the individual genders to find percent
full_gender_count = purchase_data_df["SN"].nunique()


In [82]:
# Count of Male Player
male_total = purchase_data_df[purchase_data_df["Gender"] == "Male"]["SN"].nunique()
# male_total


In [83]:
# Percent of male players
male_percent = ((male_total/full_gender_count)*100)
# male_percent

In [84]:
# Total count of Female Players
female_total = purchase_data_df[purchase_data_df["Gender"] == "Female"]["SN"].nunique()
# female_total

In [85]:
# Percent of Female Players
female_percent = ((female_total/full_gender_count)*100)
# female_percent

In [86]:
#Count of other / non-disclosed
other_total = full_gender_count - male_total - female_total
# other_total

In [87]:
# Percent of other / non-disclosed
other_percent = ((other_total/full_gender_count)*100)
# other_percent 

In [88]:
# Display the above information into a clean summary table: 
df_gender_demographics = pd.DataFrame({"Gender": ["Male", "Female", "Other / Non-Disclosed"], "Total Count": [male_total, female_total, other_total],
                                      "Percentage of Players": [male_percent, female_percent, other_percent]}, columns = 
                                     ["Gender", "Total Count", "Percentage of Players"])
gender_summary_table = df_gender_demographics.set_index("Gender")

# Round the Percentage of Players column to 2nd decimal place...refer to Stack Overlow and the rounding input...
gender_summary_table.style.format({"Percentage of Players": "{:.2f}%"})


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [127]:
#display data
purchase_data_df.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price Age Bins  
0  Extraction, Quickblade Of Trembling Hands   3.53    20-24  
1                          Frenzied Scimitar   1.56      40+  
2                               Final Critic   4.88    20-24  
3                                Blindscythe   3.27    20-24  
4                                       Fury   1.44    20-24

In [129]:
#Find the purchase count for gender
gender_purchase_data = purchase_data_df.groupby('Gender')
gender_purchase_count = gender_purchase_data['Gender'].count()

# gender_purchase_count

In [130]:
# calculate the average purchase price per gender
gender_purchase_mean = gender_purchase_data['Price'].mean()
# gender_purchase_mean

Gender
Female                   3.203009
Male                     3.017853
Other / Non-Disclosed    3.346000
Name: Price, dtype: float64

In [132]:
# Calculate the total purchase value: 
gender_purchase_total = gender_purchase_data['Price'].sum()
# gender_purchase_total

Gender
Female                    361.94
Male                     1967.64
Other / Non-Disclosed      50.19
Name: Price, dtype: float64

In [138]:
# Calulate the average price per person
avg_price_per_person = purchase_data_df.groupby(["SN", "Gender"])[["Price"]].sum().reset_index().groupby("Gender")["Price"].mean()
# avg_price_per_person

Gender
Female                   4.468395
Male                     4.065372
Other / Non-Disclosed    4.562727
Name: Price, dtype: float64

In [140]:
# Create the dataframe to display the summary table, and format to clean the data...
df_gender_pricing_analysis = pd.DataFrame({"Purchase Count": gender_purchase_count,
                                          "Average Purchase Price": gender_purchase_mean,
                                          "Total Purchase Value": gender_purchase_total,
                                           "Avg Total Purchase per Person": avg_price_per_person})
# df_gender_pricing_analysis

# Clean the data frame by making $ signs and using the 
df_gender_pricing_analysis['Average Purchase Price'] = df_gender_pricing_analysis['Average Purchase Price'].map("${:.2f}".format)
df_gender_pricing_analysis['Total Purchase Value'] = df_gender_pricing_analysis['Total Purchase Value'].map("${:.2f}".format)
df_gender_pricing_analysis['Avg Total Purchase per Person'] = df_gender_pricing_analysis['Avg Total Purchase per Person'].map("${:.2f}".format)

df_gender_pricing_analysis

Purchase Count Average Purchase Price  \
Gender                                                         
Female                            113                  $3.20   
Male                              652                  $3.02   
Other / Non-Disclosed              15                  $3.35   

                      Total Purchase Value Avg Total Purchase per Person  
Gender                                                                    
Female                             $361.94                         $4.47  
Male                              $1967.64                         $4.07  
Other / Non-Disclosed               $50.19                         $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [98]:
# Find the Max and Min values of the Age columns to find the best way to bin
print(purchase_data_df['Age'].max())
print(purchase_data_df['Age'].min())

45
7


In [99]:
# Using the table for reference, create bins in which the Data will be stored
bins = [0, 9, 14, 19, 24, 29, 34, 39, 100]

#Create names for bins
bin_group_names = ["<10", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39", "40+"]

In [100]:
# Create df for unique list of ages - one for each Screen Name 'SN' value
sn_age_group = purchase_data_df.groupby(['SN'])
df_sn_bin_age = pd.DataFrame(sn_age_group['Age'].max())

In [101]:
#Use unique age groups above and the bins/group names to place ages into bins
df_sn_bin_age["Age Bins"] = pd.cut(df_sn_bin_age["Age"], bins, labels=bin_group_names)

In [102]:
#Change index to be the age bin lables....use the inplace=TRUE - refer to STCKOVFLW
df_sn_bin_age.set_index('Age Bins', inplace=True)

In [103]:
# Assign the re arranged bins into new data frame...
# count how many values are in eacg bin by using groupby on age, bin, labels
df_bin_age = pd.DataFrame(df_sn_bin_age.groupby(['Age Bins']).count())


In [104]:
# Rename the column to 'Total Column' instead of Age...use re-name
df_bin_age.rename(columns = {'Age': 'Total Count'}, inplace = True)

In [105]:
# Calculate the % of total players in each bin
df_bin_age['Percentage of Players'] = df_bin_age['Total Count'] / total_players * 100

In [106]:
# Format the cells using map() to be 2 decimal places....refer the percent from python hw
df_bin_age['Percentage of Players'] = df_bin_age['Percentage of Players'].map("{:.2f}%".format)
df_bin_age

Total Count Percentage of Players
Age Bins                                   
<10                17                 2.95%
10-14              22                 3.82%
15-19             107                18.58%
20-24             258                44.79%
25-29              77                13.37%
30-34              52                 9.03%
35-39              31                 5.38%
40+                12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [107]:
#Bin the original dataframe by age...use the bins created above...already did that work(!)
purchase_data_df["Age Bins"] = pd.cut(purchase_data_df["Age"], bins, labels=bin_group_names)

In [118]:
# Calculation to find purchase count 
age_group = purchase_data_df.groupby(['Age Bins'])
purchase_count = age_group['Purchase ID'].count()
mean_purchase_price = age_group['Price'].mean()
total_purchase_value = age_group['Price'].sum()


In [162]:
#Summary data frame
df_purchase_analysis_age_summary = pd.DataFrame({"Purchase Count":purchase_count,
                                                "Average Purchase Price": mean_purchase_price,
                                                "Total Purchase Value": total_purchase_value,
                                               })

df_purchase_analysis_age_summary

Purchase Count  Average Purchase Price  Total Purchase Value
Age Bins                                                              
<10                   23                3.353478                 77.13
10-14                 28                2.956429                 82.78
15-19                136                3.035956                412.89
20-24                365                3.052219               1114.06
25-29                101                2.900990                293.00
30-34                 73                2.931507                214.00
35-39                 41                3.601707                147.67
40+                   13                2.941538                 38.24

I couldn't figure out the correct coding with all the bins to find the avg total purchase per person using these age groups...I will go back and fix...ran out of time.

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [24]:
# find the top 5 spenders from the list below
top_spenders = purchase_data_df.groupby('SN')

# Find total purchases and sort in descending order, find first 5 rows
top_five = top_spenders["Price"].sum().sort_values(ascending=False) [0:5]
# top_five

In [25]:
#Top spending SN index values from first rows of data...place into list
top_spending_sn = top_five.index.values[0:5]
top_spending_sn

array(['Lisosia93', 'Idastidru52', 'Chamjask73', 'Iral74', 'Iskadarya95'],
      dtype=object)

In [26]:
# Get all rows from purchase_data_df where the top five spender's SN is in "top_spending_sn"...look up .isin fx
top_five = purchase_data_df[purchase_data_df['SN'].isin(top_spending_sn)]

#group top_five by SN
top_five_group = top_five.groupby(['SN'])
top_five_group

In [27]:
# Calculate the purchase count
top_five_purchase_count = top_five_group['Purchase ID'].count().sort_values(ascending=False)
# top_five_purchase_count  

In [28]:
# Calculate the total purchases 
top_five_total_purchase = top_five_group['Price'].sum().sort_values(ascending=False)
# top_five_total_purchase

In [29]:
# Calculate the average purchase price for each top 5 sn
avg_purchase_price = top_five_group['Price'].mean().sort_values(ascending=False)
# avg_purchase_price 

In [30]:
#Display in summary table
df_top_spender_summary = pd.DataFrame({"Purchase Count": top_five_purchase_count,
                                      "Average Purchase Price": avg_purchase_price, 
                                      "Total Purchase Value": top_five_total_purchase})

#Clean the data
df_top_spender_summary['Average Purchase Price'] = df_top_spender_summary['Average Purchase Price'].map("${:.2f}".format)
df_top_spender_summary['Total Purchase Value'] = df_top_spender_summary['Total Purchase Value'].map("${:.2f}".format)

#Sort values in descending order
df_top_spender_summary.sort_values('Total Purchase Value', ascending=False)


Purchase Count Average Purchase Price Total Purchase Value
Lisosia93                 5                  $3.79               $18.96
Idastidru52               4                  $3.86               $15.45
Chamjask73                3                  $4.61               $13.83
Iral74                    4                  $3.40               $13.62
Iskadarya95               3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [19]:
# Get the Item ID, Item Name, and Item Price columns from the purchase_data_df DataFrame
df_items = purchase_data_df.loc[:,["Item ID", "Item Name", "Price"]]
# df_items.head()

In [23]:
# Group by Item ID and Item Name
item_groups = df_items.groupby(['Item ID', 'Item Name'])

#Count number of values in Price Column grouped by the Item ID
df_most_popular = pd.DataFrame(item_groups.count().sort_values('Price', ascending=False))
# df_most_popular.head()

In [32]:
# Take ITEM ID index values from the first 5 rows of data and put into a list
# The above will  be used to get all info from the top 5 most popular Item IDs
top_popular_id = df_most_popular.index.values[0:5]
# Run a For Loop to go through the Item ID column and pull out the most popular
top_popular_itemID_list = [i[0] for i in top_popular_id]
# top_popular_itemID_list

In [39]:
# Get all rows from df_items where the Item ID is in the top_popular_itemID_list
df_top_popular_id = df_items[df_items['Item ID'].isin(top_popular_itemID_list)]

# Group df_top_popular_id by Item ID and Item Name data
top_popular_group = df_top_popular_id.groupby(['Item ID', 'Item Name'])
# top_popular_group.head()

In [40]:
#Calculate the Purchase count, Item Price, and Total Purchase Value
top_popular_count = top_popular_group['Price'].count()
top_popular_price = top_popular_group['Price'].max()
top_popular_total = top_popular_group['Price'].sum()

In [43]:
#Create the summary table: 
df_top_popular_summarytable = pd.DataFrame({"Purchase Count": top_popular_count,
                                           "Item Price": top_popular_price,
                                           "Total Purchase Value": top_popular_total})

#Format to 2 decimal places to clean the data
df_top_popular_summarytable['Item Price'] = df_top_popular_summarytable['Item Price'].map("${:.2f}".format)
df_top_popular_summarytable['Total Purchase Value'] = df_top_popular_summarytable['Total Purchase Value'].map("${:.2f}".format)

# df_top_popular_summarytable

In [44]:
# Sort the values in the Total Purchase Value column in DESCENDING oder
df_top_popular_summarytable.sort_values('Purchase Count', ascending=False)

Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
108     Extraction, Quickblade Of Trembling Hands                  9   
132     Persuasion                                                 9   
145     Fiery Glass Crusader                                       9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.88   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
108     Extraction, Quickblade Of Trembling Hands         $3.53   
132     Persuasion                                        $3.33   
145     Fiery Glass Crusader                              $4.58   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
108     Extraction, Quickblade Of Trembling Hands                  $31.77  
132     Persuasion                                                 $28.99  
145     Fiery Glass Crusader                                       $41.22

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [50]:
#Count the number of values in Price column, grouped by Item ID...use item_groups from above
# Sort in DESCENDING order...that way I can find the most popular items then put into the DataFrame
df_top_profit = pd.DataFrame(item_groups.sum().sort_values('Price', ascending=False))
# df_top_profit.head()

In [53]:
# Take the ITEM ID index values from first 5 rows...like I did above
top_five_profits = df_top_profit.index.values[0:5]
top_profits_itemid_list = [i[0] for i in top_five_profits]
# top_profits_itemid_list

In [56]:
#Get all rows from df_items where the ITEM ID is in the the top profits list
df_top_five_profits = df_items[df_items['Item ID'].isin(top_profits_itemid_list)]

# Group the df_top_five_profits by Item ID and ITEM name
top_five_profit_group = df_top_five_profits.groupby(['Item ID', 'Item Name'])
# top_five_profit_group.head()

In [57]:
# Run the basic calculations for Purchase Count, Item Price, and Total Purchase Value
top_profit_count = top_five_profit_group['Price'].count()
top_profit_price = top_five_profit_group['Price'].max()
top_profit_total = top_five_profit_group['Price'].sum()

In [60]:
# Create the summary Table and include the calculations above
df_top_profit_summarytable = pd.DataFrame({"Purchase Count": top_profit_count, 
                                       "Item Price": top_profit_price, 
                                       "Total Purchase Value": top_profit_total})
# df_top_profit_summarytable

In [61]:
# Format the cells to 2 decimals with $ signs to clean data
df_top_profit_summarytable['Item Price'] = df_top_profit_summarytable['Item Price'].map("${:.2f}".format)
df_top_profit_summarytable['Total Purchase Value'] = df_top_profit_summarytable['Total Purchase Value'].map("${:.2f}".format)

#Sort the values in "Total Purchase Value" in DESCENDING order
df_top_profit_summarytable.sort_values('Total Purchase Value', ascending=False)


Purchase Count  \
Item ID Item Name                                                      
92      Final Critic                                              13   
178     Oathbreaker, Last Hope of the Breaking Storm              12   
82      Nirvana                                                    9   
145     Fiery Glass Crusader                                       9   
103     Singed Scalpel                                             8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.88   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

Three Trends from the above data set: 
1) The most profitbale and popular item is "Final Critic", with 13 purchases with a total value of $59.99, and 

2) The screen name of "Lisosia93" is the games top spender...while chamjask73 has the top average spending price per user

3) The bulk of the main player base is between the age 20-24 (around 45%)